internal/contentmanager/source/dynamicsource.go

```diff
@@ -0,0 +1,223 @@
+	// may result in attempting to configure and start an informer that has
+	// already been started.
+	DynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory
+	// GVR is the GroupVersionResource that this source is responsible
+	// for creating and configuring an informer for
+	GVR schema.GroupVersionResource
+	// Owner is the client.Object that owns the managed content that this
+	// source will be creating an informer to react to events for. This
+	// field is used to attempt to requeue the owning client.Object for
+	// reconciliation on a watch error after a previously successful sync
+	Owner client.Object
+	// Handler is the handler.EventHandler that is used to react to events
+	// received by the configured source
+	Handler handler.EventHandler
+	// Predicates are the predicate.Predicate functions used to determine
+	// if a triggered event should be reacted to
+	Predicates []predicate.Predicate
+	// OnPostSyncError is the callback function that is used when the source
+	// initially synced successfully and later encountered an error
+	OnPostSyncError func(context.Context)
+}
+
+func NewDynamicSource(cfg DynamicSourceConfig) *dynamicInformerSource {
+	return &dynamicInformerSource{
+		cfg:        cfg,
+		errChan:    make(chan error, 1),
+		syncedChan: make(chan struct{}, 1),
+	}
+}
+
+// dynamicInformerSource is an implementation of the
+// ReaderCloserSyncingSource interface. It is used
+// to create an informer, using the provided dynamic informer
+// factory, for the configured GroupVersionResource.
+//
+// The informer is configured with a WatchEventErrorHandler that
+// stops the informer, and if it had previously synced successfully
+// it attempts to requeue the provided Owner for reconciliation and
+// calls the provided OnWatchError function.
+type dynamicInformerSource struct {
+	cfg        DynamicSourceConfig
+	errOnce    sync.Once
+	errChan    chan error
+	err        error
+	syncedChan chan struct{}
+	mu         sync.Mutex
+	synced     bool
+	cancel     context.CancelFunc
+	ctx        context.Context
+	informer   informers.GenericInformer
+	started    bool
+}
+
+func (dis *dynamicInformerSource) String() string {
+	dis.mu.Lock()
+	defer dis.mu.Unlock()
+	return fmt.Sprintf("contentmanager.DynamicInformerSource - GVR: %s, Owner Type: %T, Owner Name: %s", dis.cfg.GVR, dis.cfg.Owner, dis.cfg.Owner.GetName())
+}
+
+func (dis *dynamicInformerSource) Start(ctx context.Context, q workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+	// Lock to check if this source has already been started.
+	// This should be the only time we need to lock, even if this
+	// function is called concurrently. This is because the first
+	// call that is able to lock and set the started value to
+	// true would "win" and be the one that continues on.
+	dis.mu.Lock()
+	if dis.started {
+		dis.mu.Unlock()
+		return errors.New("already started")
+	}
+
+	dis.started = true
+	dis.mu.Unlock()
```

Comment:
We need to wait to unlock until _at least_ after we set the cancel function (because of the Close implementations assumption that cancel is set if started is set)

Reply:
addressed in https://github.com/operator-framework/operator-controller/pull/1119/commits/af075c5137065a5e0b2e0cfc1b4debbb367f792f